src/modules/ekf2/EKF/mag_control.cpp

```diff
@@ -276,91 +186,76 @@ bool Ekf::resetMagStates(const Vector3f &mag)
+		_mag_decl_cov_reset = true;
 
 	} else {
 		// mag_B: reset
 		_state.mag_B.zero();
 
+		// Use the magnetometer measurement to reset the field states
 		if (reset_heading) {
-			has_realigned_yaw = resetMagHeading(mag);
+			resetMagHeading(mag);
 		}
 
 		// mag_I: use the last magnetometer measurements to reset the field states
 		_state.mag_I = _R_to_earth * mag;
 	}
 
-	resetMagCov();
+	if (!mag_I_before_reset.longerThan(0.f)) {
+		ECL_INFO("initializing mag I [%.3f, %.3f, %.3f], mag B [%.3f, %.3f, %.3f]",
+			 (double)_state.mag_I(0), (double)_state.mag_I(1), (double)_state.mag_I(2),
+			 (double)_state.mag_B(0), (double)_state.mag_B(1), (double)_state.mag_B(2)
+			);
 
-	if (has_realigned_yaw) {
-		if (!mag_I_before_reset.longerThan(0.f)) {
-			ECL_INFO("initializing mag I [%.3f, %.3f, %.3f], mag B [%.3f, %.3f, %.3f]",
-				 (double)_state.mag_I(0), (double)_state.mag_I(1), (double)_state.mag_I(2),
-				 (double)_state.mag_B(0), (double)_state.mag_B(1), (double)_state.mag_B(2)
-				);
+	} else {
+		ECL_INFO("resetting mag I [%.3f, %.3f, %.3f] -> [%.3f, %.3f, %.3f]",
+			 (double)mag_I_before_reset(0), (double)mag_I_before_reset(1), (double)mag_I_before_reset(2),
+			 (double)_state.mag_I(0), (double)_state.mag_I(1), (double)_state.mag_I(2)
+			);
 
-		} else {
-			ECL_INFO("resetting mag I [%.3f, %.3f, %.3f] -> [%.3f, %.3f, %.3f]",
-				 (double)mag_I_before_reset(0), (double)mag_I_before_reset(1), (double)mag_I_before_reset(2),
-				 (double)_state.mag_I(0), (double)_state.mag_I(1), (double)_state.mag_I(2)
+		if (mag_B_before_reset.longerThan(0.f) || _state.mag_B.longerThan(0.f)) {
+			ECL_INFO("resetting mag B [%.3f, %.3f, %.3f] -> [%.3f, %.3f, %.3f]",
+				 (double)mag_B_before_reset(0), (double)mag_B_before_reset(1), (double)mag_B_before_reset(2),
+				 (double)_state.mag_B(0), (double)_state.mag_B(1), (double)_state.mag_B(2)
 				);
 		}
-
-		_control_status.flags.yaw_align = true;
-
-		if (_control_status.flags.in_air) {
-			// record the start time for the magnetic field alignment
-			_flt_mag_align_start_time = _time_delayed_us;
-			_control_status.flags.mag_aligned_in_flight = true;
-
-		} else {
-			_flt_mag_align_start_time = 0;
-			_control_status.flags.mag_aligned_in_flight = false;
-		}
-
-		return true;
 	}
 
-	return false;
-}
-
-void Ekf::selectMagAuto()
-{
-	check3DMagFusionSuitability();
-	canUse3DMagFusion() ? startMag3DFusion() : startMagHdgFusion();
-}
+	// record the start time for the magnetic field alignment
+	if (_control_status.flags.in_air) {
+		_flt_mag_align_start_time = _time_delayed_us;
+		_control_status.flags.mag_aligned_in_flight = true;
 
-void Ekf::check3DMagFusionSuitability()
-{
-	checkYawAngleObservability();
-	checkMagBiasObservability();
-
-	if (_mag_bias_observable || _yaw_angle_observable) {
-		_time_last_mov_3d_mag_suitable = _time_delayed_us;
+	} else {
+		_flt_mag_align_start_time = 0;
+		_control_status.flags.mag_aligned_in_flight = false;
```

Comment:
rather let the reset to be done in `controlMagFusion` only?

Reply:
Yes that's better, updated.